cmd/prometheus/main_test.go

```diff
@@ -155,3 +165,54 @@ func TestComputeExternalURL(t *testing.T) {
 		}
 	}
 }
+
+// TestApplyConfig ensures that no component is modifing the global config.
+// The config includes many pointers so modifing it has global side effects.
+func TestApplyConfig(t *testing.T) {
+	// Disable k8s logs.
+	k8s_runtime.ErrorHandlers = []func(error){}
+
+	var (
+		cfgOrig          *config.Config
+		ctx, cancel      = context.WithCancel(context.Background())
+		logger           = log.NewNopLogger()
+		notifyManager    = notifier.NewManager(&notifier.Options{}, logger)
+		discoveryManager = discovery.NewManager(ctx, logger)
+		scrapeManager    = scrape.NewManager(logger, nil)
+		webHandler       = web.New(logger, &web.Options{RoutePrefix: "/"})
+		remoteStorage    = remote.NewStorage(logger, nil, time.Duration(1*time.Millisecond))
+	)
+	defer cancel() // Just in case to avoid goroutine leaks.
+
+	// Load the config in all components.
+	reloaders := []func(cfg *config.Config) error{
+		// Save the config so we can compare it with a new untouched copy.
+		// Since this is a pointer any modification will be reflected in the saved copy.
+		func(cfg *config.Config) error {
+			cfgOrig = cfg
+			return nil
+		},
+		remoteStorage.ApplyConfig,
+		webHandler.ApplyConfig,
+		notifyManager.ApplyConfig,
+		scrapeManager.ApplyConfig,
+		scrapeManager.ApplyConfig,
+		func(cfg *config.Config) error {
+			c := make(map[string]sd_config.ServiceDiscoveryConfig)
+			for _, v := range cfg.ScrapeConfigs {
+				c[v.JobName] = v.ServiceDiscoveryConfig
+			}
+			return discoveryManager.ApplyConfig(c)
+		},
+	}
+	testutil.Ok(t, reloadConfig("../../config/testdata/conf.good.yml", logger, reloaders...))
```

Comment:
There is not a test case in this file that covers the test you are removing.

Reply:
not sure what you mean here. 
If you revert the changes for this bug https://github.com/prometheus/prometheus/issues/4214 you will see that the tests will fail.

the bug is triggered by the following execution

config reload is triggered.
scrape.ApplyConfig - loads and saves the config.
discovery.ApplyConfig - changes the config and since the some fields are pointers the saved config in scrape.ApplyConfig is also changed.

another config reload is triggered without any config file changes.
scrape.ApplyConfig - checks the provided config to check if it matches with the saved one and since the config was changed by discovery.ApplyConfig the check fails and triggers scrape loop reload.

this test ensures that no ApplyConfig modifies the original file which will prevent such bugs.

Reply:
good.yml doesn't contain a static_configs with multiple entires.

Reply:
that is not important in this case.

Reply:
actually I just double checked and it does have a static config with 2 targets in it.

https://github.com/prometheus/prometheus/blob/91ce63a1402ff2e2447eb0286ba0e0c17b64624a/config/testdata/conf.good.yml#L50-L54

still trying to figure out the failing race

Reply:
That's only one static_config though

Reply:
Adding more to this file brakes `TestLoadConfig` and I really don't see why you think it is necessary. the bug wasn't related at all to having more than one static config entry.
if you really think this is necessary lets do it in a separate PR as it will require a non related changes to the `TestLoadConfig` if we do it here.

Reply:
We still need a test added with two static_configs to ensure this doesn't regress.

Reply:
I understand your point, but as I mentioned this PR solves a different problem so your concern needs to be addressed in a separate one.

Reply:
You are still removing test cases without replacing them.

Reply:
the test I am removing is completely covered in by the change I am suggesting.

The test I am removing is so specific that it wouldn't cover regressions anyway.

Reply:
the original bug was not about this, but it seems that we won't come to an agreement this way so I updated the test to use a different config file. Hopefully this should satisfy your concern.

Reply:
Thanks, there's no need to create a new file though. The existing good configuration is meant to be an exhaustive example of every possible configuration option, so you can add it in there.

Reply:
I already mentioned that I did this, but it brakes some other tests and I don't feel like fixing unrelated tests. this should be done in a separate PR.